Refactor by using an array of directives

joker314 opened the pull request:
Resolves #38 

The array of directives is at the top, and can easily be modified if any new directives are added.

joker314 committed:
refactor: use an array of directives

joker314 commented:
Of course, one still needs to specify what you're allowed to put inside a directive in the `validate` method.
<hr>

I add a (public) method `camelCase`. Should I make it more private? If so, how?

joker314 committed:
style: lint

joker314 commented:
Lint is failing due to a security issue. It claims that `a[b]` notation is bad: but it appears to be a false positive, since `b` in our case is one of the values in the array at the top of the file (i.e., we have full control over it) so it can't hijack the prototype chain or anything like that.

Additionally, we do a `hasOwnProperty` check, so `a[b]` will not be accessing the prototype at all.

I'm going to disable the lint for that line or find a 'safer' way of writing it down.

joker314 committed:
style: Disable security test temporarily

Normally, testing for security issues is very important. However, in this
particular instance, on this particular line, we have determined that
there is no threat to security. Hence, we disable the check.

We have come to this conclusion by understanding that the relevant
security issue occurs when we allow unfiltered user input as the key to an
object. As JavaScript traverses the prototype chain, certain unexpected
values could be accessed.

This was a non-issue because we do a .hasOwnProperty check on the key we
intend to use. This means that the prototype chain will not be traversed
because a match will be found instantly (we use 'continue' to escape).

Additionally, the keys we attempt to use are already safe: we pick them
from a whitelist mapped with a pure function. When we apply this pure
function to each element in our whitelist, we don't (to the best of our
knowledge) get anything which may be harmful, such as 'constructor'. This
may become an issue in future if a 'Constructor' directive is added, but
given the previous defence this is still a non-issue.

When disabling lint, we do not perform a blanket disable. Instead, we
target a specific line for this specific security issue. If a different
security issue is introduced in future on that line, or this security
issue is introduced on a different line, it will be detected, hence this
change does not compromise security.

lirantal commented on index.js:
```diff
@@ -105,6 +87,20 @@ class middleware {
+   * Converts a security.txt directive like 'Contact'
+   * to the camelCase field name like 'contact'
+   *
+   * We assume that the contains a sequence of capitalised
```
"We assume that the contains" ?

lirantal commented:
LGTM

joker314 committed:
docs: Insert missing word into code comment

We also replace 'contains' with 'is' since the former implies we expect leading and trailing garbage

joker314 commented:
Nice catch! I fixed the typo; but for me to be allowed to merge it I need you to rereview.

joker314 committed:
test(formatPolicy.test.js): Test camelCase utility method to increase coverage

When converting a string to camelCase, there's a special case: the first word must be in lowercase,
but subsequent words must be capitalised. The specification currently only includes single words,
but there are proposals for multiple words seperated by dashes (and we support this if these are
accepted). The other tests therefore were only hitting the special case, and so part of the ternary
condition is untested. Coverage doesn't like this, so we add three tests for non-existant
directives: Abc, Abc-Def, Abc-Def-Ghi; and these restore coverage to 100%

joker314 committed:
Merge branch 'refactor' of https://github.com/lirantal/express-security-txt into refactor

joker314 commented:
Also: increased coverage to 100%. Ready for re-review.

lirantal commented:
:tada: This PR is included in version 2.2.3 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/express-security-txt)
- [GitHub release](https://github.com/lirantal/express-security-txt/releases/tag/v2.2.3)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
